### ISSUE TITLE
show_opt_in property does not work as expected ?

### ISSUE THREAD

**Issue by Cheprod** [json-editor/json-editor#505]
### General information

- json-editor version: 1.4.0-beta.0

### Expected behavior

I want to use show_opt_in set to true to display all defined field in the jsonschema (non required field included).  For non-required fields : 
- If the property is present in the json, display the field with the value + a checkbox checked to delete the property from the json if the user want.
- If the property isn't present in the json, display a blank field + a checkbox not checked to add the property to the json. If defaultValue on the schema is defined, I want to fill the field with the default value from the jsonschema.

### Actual behavior

When the property is present in the json, the field is displayed but the checkbox isn't checked and field isn't editable. 

When the property isn't present in the json, the field is not displayed at all.

### Steps to reproduce the behavior

I can't reproduce the behaviour on online editor but I share my json and my schema.

Direct link to example: https://json-editor.github.io/json-editor/?schema=N4IgLglmA2CmIC4QDMJwCIEMyYEqwGMB7AJwBNYSQAaFUgW20RAO1gHNSJYBnGkAEaYeEAgGE2nEgE8AKlDjMACiQgA7AhAAOmaPwo8CqrZCJrmYs6nYBXEtghmABAGUCAC1iMnyUk4BiaLBYOPjE5JT8YNJa8EhEAgBWhGD8WiREsSSQvIigBkbapuYI+byFJo4lIOjlxsVOZDZO9ERNirTRscw8YKpq7CAAvrQ6YO55IAX1VRae9OpOJLDQDshORACfTnA+ou7cJE48lJhOAJeFYAB0TgCysGB9sE4AOlPYmO9OOkdk58hMDZUp0YnEQL1+oNaBRATZoKkkGRPsNaKg4EpsGBKCUyoYZmZlFicY1mmoiPRSXsPIdGi8yGY1OdeBttnAeE4AG66WB2DnEej0WAaWCYNRgH6YI4AUh+vM5wolACvnksvFpVgRmRyABScogQI6rHxmAgNLTScbOSHISBCgCUt0xR1gAA91XBqE5pQBNaX0aVkWTSgAS0ru0pcAH07mZAjTKNcCDxOU52Oc1JRTk4bGonOTKRRqQdKFSui8AEwABgAjABOKsADgrAHYALRNisAVirVZjcf2hyTKaiYJ6fXU0KmsDhCOYvrb/sXQdDi7ui+jsbU8ZLJGHnNRIAyOGxmKeOMm0yKsyQZ+xJDzRePDmcCB2pB4AEcbC9oABydR73Sc4X0fZoxnvPMPDFdhYC9XYCypVAEyOYg7DFCUiwZNQmRZE57DVVhoAIeFzlHbokEhSd+GfU9iQfecfUXehl1kQ8CGkAg4H8ZZv2FDjLzqa9CSQWR1Q5IsOK4+lmladpYFuAAZTAIB4fDJT+AEgQlGsq2OQgzAMa5yPBNQbHoARIloBY1AgehzMQKsYRnIE5wQXSRghWBfmwUhBPxYTqhcc4fOxOwnEVZYdhZaD6C0CTmmQ3cnDEFwADVjNBCiIQnAZ9Bc+FERAABuQ9kGgGweHcO5MFdeQhX8ipimYMS4o5RhXTs8yIvFKKKx8CqqqQwdS2FfTiDUIyTOYMyLKskAbK6+hHOc2dEQ8tFBuq2rFPUOI8Sam8QAAOQpAQoo6pa6R2CB2EzDlLlUbEOUwblxWigbKvcYaUMy8AxyQWbLKoaz1CWlbpzWxANpASrYAAVVkMRGoJaoAElJogPjjggDY82BNBVN/P9PDsF5EbEH4iHC3Y7V4HA4umpABCIIg4DFfKoYQPof08uGKcCOBjswBrSmnALKhEkAMbILGfxxvGc0gaAiZ2EneSiimqZpl4C3oc4PqSw4mcEVn2fMVbXMRHnYE8unehFrR/AYJgxavSXqmdkhGEw5p7YZrQnB1N0nDfQMAHp/Ujn1vRDBBw3jlxHScZTVJ4A3OQgIstGpo4CZV9Tdl8b3sA0xotOBIOvZ9pw0ZcAB5RsADZa0dE2qLyy3CuYQ8cHYHhhdFg7UeYXbenpFkCwkl4+45AADi5Cii5E1A5IuRr+elGW1W4XFxoESLzWf8z/enji0S4IFQc4vTQDl8LMCUwGph6rgksU15eI3ExNqV7GkfgUAvB8DFuWccUJDw2TRtiegICnKw1snxaBwDEA208gAa1gN5QWvB0B2EnBjFwhAUaBWYKdCyy8XgnAmgYOUk0MLRS/D+aA/IzD4QVEaFk38SDT0aNvXgSkVJqVLAvPSWQhSYV1i8HQ98bBaHSHZUs7JiyHB4H9MBSBAIcHmoteyy0EBVk8gKBYYAXC5wILAQe+1xaHSluQpwAoNSPCpHJeE9I/xbB+BkTOioPrGKgO3XKgwhieV4jYQ0sAyCIAANo2JHqMbAEw0RBDvBeWgtFYCpIYrQE4oU/KbS+jVOqij+DlUKTtPa/B+ZIyqScAWQQrFREwP3RptBMHYKCDwPBUJCHENoP40x5jLEi3gAAXSGEAA===&value=N4IgJgpgzgxgTgSwA4BcEHsB2IBcIQA0ISAhigBa7hkmEgBmCANhAApkoRzZ4CkAmgFpeAW2FgAKrwASwgLLCAygH05WAGIIY5BFwB0MKADc6cdCjJsOXHviIwAnjBbq4EAI4BXCJke4AjAAMRFAQpHBk6HBUANx09EyeUORyJAAeEggiEAHBDInJqWkAMgiYOThBREkQAKoSAMK4KHDe1aH1DZosAHIk2c2tEEQWAOZQfQM4ANoAukQA1hBh3dAAIp6ImKMAkpiKEDC4eTDoIiIIKIromzAQkxX4AL5AA==&lib_switcher=&prompt_before_delete&upload=function(a,b,c){console.log(%22Upload%20handler%20required%20for%20upload%20editor%22)}&theme=foundation5&iconlib=fontawesome4&object_layout=grid&show_errors=interaction&required_by_default

Extract from my code :

```javascript
JSONEditor.defaults.editors.object.options.collapsed = false;
JSONEditor.defaults.options.show_opt_in = true;

this.editor = new JSONEditor(element, {
					  theme: 'foundation5',
					  schema: schemaGlobal,
					  iconlib: "fontawesome4",
					  disable_collapse: false,
					  disable_edit_json: true,
					  disable_properties: true,
					  enable_array_copy: true,
					  startval: jsonData
					});
```
Thanks a lot
Regards

**Comment by Cheprod** [comment]
Looking at the code and doing some tries :

- It's not necessary to use the required_by_default property set to true
- Nothing is done in the code to tick the checkbox when a non required properties has a value in the json
- The method deactivateNonRequiredProperties hide the non required properties that hasn't a value in the json

**Comment by Cheprod** [comment]
For short term solution, I just add this hack in the setValue method :

```
  setValue: function(value, initial) {
    var self = this;
    value = value || {};

    if(typeof value !== "object" || Array.isArray(value)) value = {};

    // First, set the values for all of the defined properties
    $each(this.cached_editors, function(i,editor) {
      // Value explicitly set
      if(typeof value[i] !== "undefined") {
        self.addObjectProperty(i);
        editor.setValue(value[i],initial);
        // TRO Hack pour ne pas masquer les paramètres non valorisés
        if (!self.isRequired(editor) && JSONEditor.defaults.options.show_opt_in) {
          editor.activate();
        }
      }
      // Otherwise, remove value unless this is the initial set or it's required
      // TRO Hack pour ne pas masquer les paramètres non valorisés
      else if(!initial && !self.isRequired(editor) && !JSONEditor.defaults.options.show_opt_in) {
        self.removeObjectProperty(i);
      }
      // Otherwise, set the value to the default
      else {
        editor.setValue(editor.getDefault(),initial);
      }
    });

    $each(value, function(i,val) {
      if(!self.cached_editors[i]) {
        self.addObjectProperty(i);
        if(self.editors[i]) self.editors[i].setValue(val,initial);
      }
    });

    this.refreshValue();
    this.layoutEditors();
    this.onChange();
  }
```

**Comment by germanbisurgi** [comment]
hi Cheprod. I didn´t really understand the problem but i´ll try to crack this peace by peace.

> "I want to use show_opt_in set to true to display all defined field in the jsonschema (non required field included)"

show_opt_in: if true, NON required properties will have an extra toggable checkbox near the title that determines if the value must be included or not in the editor´s value. This apply to defined and not defined property values.

> "For non-required fields :
> If the property is present in the json, display the field with the value + a checkbox checked to delete the property from the json if the user want."

This is how it works already. The non required field will have the opt_in checkbox. If is not checked the value will not be present in the final output.

> "For non-required fields :
> If the property isn't present in the json, display a blank field + a checkbox not checked to add the property to the json. If defaultValue on the schema is defined, I want to fill the field with the default value from the jsonschema."

if the property isn´t present in the json (schema?) then there is no editor. 

> "For non-required fields :
> If defaultValue on the schema is defined, I want to fill the field with the default value from the jsonschema."

It does already

**Comment by Cheprod** [comment]
Hi germanbisurgi,

Thanks for your help!

I did an simple example to illustrate what I said (I didn't succeed with the online editor because the show_opt_in option isn't available). This is done with the 1.4.0_beta0 version.

This is my JSON schema :
```
{
	"id" : "confAcquisitionRandom",
	"title" : "Configuration Acquisition Random",
	"options": {
		"disable_collapse" : true
	},
	"format": "categories",
	"basicCategoryTitle": "Principal",
	"type" : "object",
	"description" : "Configuration Acquisition Random",
	"properties" : {
		"description" : 	{
			"description": "Description of the module",
			"type" : "string"
		},
		"cycleFrequency" : {
			"description" : "Temps entre 2 cycles",
			"type" : "number",
			"default" : 10,
			"minimum" : 0
		},
		"warningWhenErrorsIn" : {
			"description" : "Warning when an error occured in the last nb seconds",
			"type" : "number",
			"default" : 600,
			"minimum" : 0
		}
	},
	"required": ["description"] 
}
```
This is my JSON data :

```
{
    "cycleFrequency": 10, 
    "description": "Acquisition Random"
}
```

Note that the property "warningWhenErrorsIn" isn't present in the json.

What I have in my editing screen :

![CC01](https://user-images.githubusercontent.com/36696593/64432705-5aaa0900-d0bd-11e9-84ba-837be1f5f9cf.jpg)

What I would have (just did a little hack on the code of the 1.4.0 beta 0, code above) :

![CC01_hack](https://user-images.githubusercontent.com/36696593/64432753-70b7c980-d0bd-11e9-8f32-6a01439bb9cd.jpg)

The last option I want have is to hide the default value when the checkbox isn't ticked (to avoid user misunderstanding). So, when the user tick the checkbox, if no value is available, the default value is put in the field.

Is it clear or not ?

Maybe the default behaviour of the json editor isn't adapted to my request.

**Comment by germanbisurgi** [comment]
yes i see better :)

i think you are right. this are my thoughts:
- the initial form build is correct
- setting a required value from JSON value works as expected
- setting a no required value from JSON value still displays the editor and the checkbox as expected but the field should remain disabled until the checkbox is checked (must to be fixed).  
- omitting a non required value in the JSON when save is clicked should still display the non required editor with a checkbox and this is not the case (must to be fixed)
- the extra feature could be added, maybe through a configuration setting like "hide_empty_optional"

**Comment by pmk65** [comment]
@germanbisurgi Is this fixed in the 2.x.x version? If so we can close this.

**Comment by germanbisurgi** [comment]
fixed in #943
